lib/spack/spack/cmd/config.py

```diff
@@ -135,11 +164,150 @@ def config_list(args):
+    if args.file:
+        # Get file as config dict
+        with open(args.file, 'r') as f:
+            data = syaml.load_config(f)
+
+        if 'env' in data or 'spack' in data:
+            # Validate as an environment
+            # use config_dict to remove top-level key
+            spack.environment.validate(data, 'spack.yaml')
+            config_dict = spack.environment.config_dict(data)
+        else:
+            # Separate config file
+            # Use schema according to first key (they each only accept one key)
+            # Do not remove top key
+            spack.config.validate(
+                data, spack.config.section_schemas[next(iter(data))])
+            config_dict = data
+        # update all sections from config dict
+        # We have to iterate on keys to keep overrides from the file
+        for section in config_dict.keys():
+            if section in spack.config.section_schemas.keys():
+                # Special handling for compiler scope difference
+                # Has to be handled after we choose a section
+                if scope is None:
+                    if section == 'compilers':
+                        scope = spack.config.default_modify_scope()
+                    else:
+                        scope = spack.config.default_modify_scope(False)
+
+                value = config_dict[section]
+                existing = spack.config.get(section, scope=scope)
+                new = spack.config.merge_yaml(existing, value)
+
+                if re.match(r'env.*', scope or ''):
+                    e = ev.get_env(args, 'config add')
+                    e.set_config(section, new)
+                else:
+                    spack.config.set(section, new, scope=scope)
+
+    if args.value:
+        components = spack.config.process_config_path(args.value)
+
+        has_existing_value = True
+        path = ''
+        override = False
+        for idx, name in enumerate(components[:-1]):
+            # First handle double colons in constructing path
+            colon = '::' if override else ':' if path else ''
+            path += colon + name
+            if getattr(name, 'override', False):
+                override = True
+            else:
+                override = False
+
+            # Test whether there is an existing value at this level
+            existing = spack.config.get(path, scope=scope)
+
+            if existing is None:
+                has_existing_value = False
+                # We've nested further than existing config, so we need the
+                # type information for validation to know how to handle bare
+                # values appended to lists.
+                existing = spack.config.type_of(path)
+
+                # construct value from this point down
+                value = syaml.load_config(components[-1])
+                for component in reversed(components[idx + 1:-1]):
+                    value = {component: value}
+                break
```

Comment:
can the logic above go somewhere in `spack.config`?  This is a lot of detailed knowledge for a command to have about internals.

Reply:
I tend to agree with @tgamblin that this should probably go in `spack.config`. I don't think though that we have to do this change necessarily in this PR, since it can be part of a following PR shaped in a way that it would be transparent to users.